File: hangman.c
#include <stdio.h>

int complete(int alpha[]);

int main(){
    //Standard in
   FILE *fin;
   fin = stdin;
   
   int times, k;
   
   //Read in the number of lava pits to process.
   fscanf(fin, "%d\n", &times);
   
   for(k = 1; k <= times; k++){
      int feet, i;
      char password[101]; //Will store the password string
      char guesses[101]; //Will store the guessing string
      
      //Read case input
      fscanf(fin, "%d %s %s", &feet, password, guesses);
      
      //Declare and initialize a boolean array corresponding to each letter in the alphabet.
      int alpha[26]; 
      for(i = 0; i < 26; i++)
         alpha[i] = 0;
      
      //If a letter appears in password, we'll set it to true. This will allow us to keep track of Ali's guessing progress.  
      for(i = 0; i < 101 && password[i] != '\0'; i++)
         alpha[password[i]-'a'] = 1;
      
      i = 0; //From here on, i will iterate over guesses
      
      //The game only ends when we run out of feet or the password is guessed completely
      while(feet > 0 && !complete(alpha)){
         //For each character in guesses, it will either be a hit or miss
         if(alpha[guesses[i]-'a']){ //If the character is set to true, it's a hit! We set it to false and continue looping.
            alpha[guesses[i]-'a'] = 0;
         }else{ //Otherwise, we've either seen the character before or it doesn't exist. Either way, it's a miss. We lower Ali one foot closer to the lava.
            feet--;
         }
         i++; //Move to the next guess
      }
      
      //If we've completed the password, Ali survives! Otherwise, poor Ali.
      if(complete(alpha)){
         printf("Lava pit #%d: Ali survives!\n",k);
      }else{
         printf("Lava pit #%d: Poor Ali.\n",k);
      }
   }

   return 0;
}

//Iterates through all remaining hidden letters of the password, returning false if any remain. Otherwise, returns true.
int complete(int alpha[]){
   int i;
   for(i = 0; i < 26; i++)
      if(alpha[i])
         return 0;
   return 1;
}

File: hangman.java
import java.util.*;
import java.io.*;

public class hangman {
	public static void main(String[] args) throws IOException {
	
		//Initialize the scanner
		Scanner in = new Scanner(System.in);
		
		//Read in the number of games to be checked
		int T = in.nextInt();
		
		//Loop through each game
		for (int t=0; t<T; t++) {
		
			//Input the starting height
			int N = in.nextInt();
			
			//Input the key
			String key = in.next();
			
			//Input the guesses
			String guesses = in.next();

			//Initialize variables			
			int uniqueChars = 0; //The number of unique characters in the key
			boolean[] hasChar = new boolean['z'-'a'+1]; //If the character exists in the key (note: the quantity of each letter does not matter)
			
			//Loop through the key, and set the correct values for uniqueChars and hasChar
			for (int k=0; k<key.length(); k++)
				if (!hasChar[key.charAt(k)-'a']) { //If the character has not been seen yet
					hasChar[key.charAt(k)-'a'] = true; //It has now been seen; it exists in the key
					uniqueChars++; //This is another unique character
				}
			
			//Loop through each letter of guess (while there are still characters to guess, and there 
			// are more needed to be guessed, and he is not in the lava)
			for (int g=0; g<guesses.length() && uniqueChars > 0 && N > 0; g++) {
				if (!hasChar[guesses.charAt(g)-'a']) //If the character does not exist in the key
					N--; //Lower poor Ali's height by 1
				else { //If the chracter does exist in the key
					uniqueChars--; //There is now one less unique character needed to be guessed
					hasChar[guesses.charAt(g)-'a'] = false; //But all occurances of that character are now gone from the key
				}
			}
			
			if (N > 0) //If Ali is done, and is above the lava, YAY!
				System.out.printf("Lava pit #%d: Ali survives!\n",t+1);
			else //If he is in or below the lava, poor Ali, I guess he won't be coming back next year
				System.out.printf("Lava pit #%d: Poor Ali.\n",t+1);
			
		}
	}
}
